analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/ControllersHaveMixedResponsibilities.Latest.cs

```diff
@@ -0,0 +1,1139 @@
+}
+
+[ApiController]
+public class WithMethodsDependingOnEachOther : ControllerBase // Noncompliant
+{
+    IS1 s1; IS2 s2; IS3 s3; IS4 s4; IS5 s5; IS6 s6; IS7 s7;
+
+    // Chain: A2 to A1
+    void A1() { s1.Use(); }            // Secondary {{May belong to responsibility #1.}}
+    void A2() { s2.Use(); A1(); }      // Secondary {{May belong to responsibility #1.}}
+    void A3() { s2.Use(); }            // Secondary {{May belong to responsibility #1.}}
+    // 1-cycle A4 => no service used => ignore
+    void A4() { A4(); }
+    // 2-cycle A5, A6 => no service used => ignore
+    void A5() { A6(); }
+    void A6() { A5(); }
+    // 3-cycle A7, A8, A9 => no service used => ignore
+    void A7() { A8(); }
+    void A8() { A9(); }
+    void A9() { A7(); }
+    // 3-cycle A10, A11, A12 with A13 depending on A12 via s3
+    void A10() { A11(); }              // Secondary {{May belong to responsibility #2.}}
+    void A11() { A12(); }              // Secondary {{May belong to responsibility #2.}}
+    void A12() { A10(); s3.Use(); }    // Secondary {{May belong to responsibility #2.}}
+    void A13() { s3.Use(); }           // Secondary {{May belong to responsibility #2.}}
+    // 3-cycle A14, A15, A16 with chain A18 -> A15 via s4
+    void A14() { A15(); s4.Use(); }    // Secondary {{May belong to responsibility #3.}}
+    void A15() { A16(); }              // Secondary {{May belong to responsibility #3.}}
+    void A16() { A14(); }              // Secondary {{May belong to responsibility #3.}}
+    void A17() { s4.Use(); }           // Secondary {{May belong to responsibility #3.}}
+    // Independent method => no service used => ignore
+    void A18() { }
+    // Independent method with its own service
+    void A19() { s5.Use(); }           // Secondary {{May belong to responsibility #4.}}
+    // Two actions calling a third one
+    void A20() { A22(); }              // Secondary {{May belong to responsibility #5.}}
+    void A21() { A22(); }              // Secondary {{May belong to responsibility #5.}}
+    void A22() { s6.Use(); s7.Use(); } // Secondary {{May belong to responsibility #5.}}
+}
+
+[ApiController]
+public class WithServiceProvidersInjectionCoupled : ControllerBase // Compliant: s1Provider and s2Provider are known to provide services
+{
+    private readonly Func<IS1> s1Provider;
+    private readonly Func<int, IS2> s2Provider;
+
+    public WithServiceProvidersInjectionCoupled(Func<IS1> s1Provider, Func<int, IS2> s2Provider)
+    {
+        this.s1Provider = s1Provider;
+        this.s2Provider = s2Provider;
+    }
+
+    public void A1() { s1Provider().Use(); var s2 = s2Provider(42); s2.Use(); }
+    public void A2() { (s2Provider(42)).Use(); }
+}
+
+[ApiController]
+public class ApiController : ControllerBase { }
+
+public class DoesNotInheritDirectlyFromControllerBase(IS1 s1, IS2 s2) : ApiController // Compliant, we report only in classes that inherit directly from ControllerBase
```

Comment:
I would add a test that looks exactly like this, but derives from `Controller`.
Just to be explicit that we do not take it into account.